dbms/src/DataStreams/ScanHashMapAfterProbBlockInputStream.cpp

```diff
@@ -74,25 +77,61 @@ struct AdderNonJoined<ASTTableJoin::Strictness::All, Mapped>
     }
 };
 
-NonJoinedBlockInputStream::NonJoinedBlockInputStream(const Join & parent_, const Block & left_sample_block, size_t index_, size_t step_, size_t max_block_size_)
+template <bool add_joined, typename Mapped>
+struct AdderMapEntryRightSemi
+{
+    static size_t add(const Mapped & mapped, size_t key_num, size_t num_columns_left, MutableColumns & columns_left, size_t num_columns_right, MutableColumns & columns_right, const void *& next_element_in_row_list, const size_t max_row_added)
+    {
+        size_t rows_added = 0;
+        auto current = &static_cast<const typename Mapped::Base_t &>(mapped);
+        if unlikely (next_element_in_row_list != nullptr)
+            current = reinterpret_cast<const typename Mapped::Base_t *>(next_element_in_row_list);
+        for (; rows_added < max_row_added && current != nullptr; current = current->next)
+        {
+            bool flag = current->getUsed();
+            if constexpr (!add_joined)
+                flag = !flag;
+            if (flag)
+            {
+                /// handle left columns later to utilize insertManyDefaults
+                for (size_t j = 0; j < num_columns_right; ++j)
+                    columns_right[j]->insertFrom(*current->block->getByPosition(key_num + j).column.get(), current->row_num);
+                ++rows_added;
+            }
+        }
+        for (size_t j = 0; j < num_columns_left; ++j)
+            /// should fill the key column with key columns from right block
+            /// but we don't care about the key column now so just insert a default value is ok.
+            /// refer to https://github.com/pingcap/tiflash/blob/v6.5.0/dbms/src/Flash/Coprocessor/DAGExpressionAnalyzer.cpp#L953
+            /// for detailed explanation
```

Comment:
These comments does not apply to rightsemi? I think we can add default value here because right semi does not output any left columns?

Reply:
And looks like `num_columns_left ` for right semi join?

Reply:
Yeah, if change the struct name to a more general one like "AdderRowFlaggedMapEntry", then num_columns_left could be non-zero, and could be used by other join types.